opt: panic: provided does not intersect required

Caught by nightly logic test suite: https://teamcity.cockroachdb.com/viewLog.html?buildId=1019282&buildTypeId=Cockroach_SqlLogicTests&tab=buildLog&_focus=14082

```
[Test Output]
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_24.test: 3432 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_22.test: 1086 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_5.test: 2345 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/100/slt_good_0.test: 9076 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/10/slt_good_1.test: 7248 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_23.test: 6780 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_9.test: 9787 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/between/1/slt_good_0.test: 2032 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_24.test: 3968 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_22.test: 1737 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_5.test: 2408 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/100/slt_good_0.test: 9094 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/10/slt_good_1.test: 8042 statements/queries
--- done: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_9.test with config local-opt: 10034 tests, 0 failures
--- total progress: 40279 statements/queries
=== CONT  TestSqlLiteLogic/local-opt/slt_good_21.test#02
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_23.test: 7522 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/between/1/slt_good_0.test: 2057 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_24.test: 4286 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_22.test: 1781 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_5.test: 2507 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/100/slt_good_0.test: 9112 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/10/slt_good_1.test: 8114 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_23.test: 8159 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_21.test: 159 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/between/1/slt_good_0.test: 2103 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_24.test: 5106 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_22.test: 2410 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_5.test: 2555 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/100/slt_good_0.test: 9196 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby/10/slt_good_1.test: 8769 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_23.test: 8921 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/orderby_nosort/10/slt_good_21.test: 192 statements/queries
panic: provided does not intersect required [recovered]
	panic: panic while executing 1 statements: SELECT _, _ FROM _ WHERE ((((((_ > _) AND (_ > _)) AND (_ <= _)) AND (_ < _)) AND (_ IN (SELECT _ FROM _ WHERE ((((_ <= _) AND ((_ BETWEEN _ AND _) OR ((_ > _) OR (_ > _)))) AND (_ >= _)) AND (_ > _)) AND (_ > _))))) ORDER BY _ DESC; caused by provided does not intersect required

goroutine 82898 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc424112000, 0x2afc9c0, 0xc424a17380, 0x223b560, 0x2ac9650)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:647 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc424112000, 0x2afc9c0, 0xc424a17380)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:388 +0x61
panic(0x223b560, 0x2ac9650)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/ordering.trimProvided(0x0, 0x0, 0x0, 0xc4279ef858, 0xc42e478848, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/ordering/ordering.go:264 +0x1d7
github.com/cockroachdb/cockroach/pkg/sql/opt/ordering.mergeJoinBuildProvided(0x2b26620, 0xc4262beb60, 0xc4279ef858, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/ordering/merge_join.go:81 +0xd3
github.com/cockroachdb/cockroach/pkg/sql/opt/ordering.BuildProvided(0x2b26620, 0xc4262beb60, 0xc4279ef858, 0x3c5ba01, 0x2b07480, 0xc4262beb80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/ordering/ordering.go:71 +0x82
github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).setLowestCostTree(0xc424112f80, 0x2b07d40, 0xc4262beb60, 0xc4279ef840, 0xc4279ef7c0, 0xc4279ef840)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:655 +0x2fa
[01:59:17][Test Output]
github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).setLowestCostTree(0xc424112f80, 0x2b08200, 0xc42e478b68, 0xc4279ef7c0, 0xc42e478b68, 0x1b541d8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:642 +0x13f
github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize(0xc424112f80, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:200 +0x11f
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc424112428, 0x2afca80, 0xc42dd53fb0, 0x2b00640, 0xc4279ef640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:464 +0x8e9
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optionallyUseOptimizer(0xc424112428, 0x2afca80, 0xc42dd53fb0, 0x0, 0x0, 0xc4245e3011, 0x4, 0x0, 0x3, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/planner.go:546 +0xce
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc424112000, 0x2afca80, 0xc42dd53fb0, 0x2b00640, 0xc4279ef640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:781 +0x18e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc424112000, 0x2afca80, 0xc42dd53fb0, 0x2b00640, 0xc4279ef640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:416 +0xe7a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc424112000, 0x2afca80, 0xc42dd53fb0, 0x2b00640, 0xc4279ef640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99 +0x341
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc424112000, 0x2afc9c0, 0xc424a17380, 0xc428bfb2d8, 0x5400, 0x15000, 0xc428bfb370, 0xc423296470, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1113 +0x219b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc425e6fe00, 0x2afc9c0, 0xc424a17380, 0xc424112000, 0x5400, 0x15000, 0xc428bfb370, 0xc423296470, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:390 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc425e6fe00, 0x2afc9c0, 0xc424a17380, 0xc424112000, 0x5400, 0x15000, 0xc428bfb370, 0xc423296470, 0xc423296480, 0xc42421c840)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:309 +0x81
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:308 +0x107c

goroutine 1 [chan receive, 1 minutes]:
testing.(*T).Run(0xc4201e3680, 0x25e973d, 0x10, 0x2708360, 0x8ba501)
	/usr/local/go/src/testing/testing.go:825 +0x301
testing.runTests.func1(0xc4201e3590)
	/usr/local/go/src/testing/testing.go:1063 +0x64
testing.tRunner(0xc4201e3590, 0xc420265dd8)
	/usr/local/go/src/testing/testing.go:777 +0xd0
testing.runTests(0xc42034a620, 0x3a022a0, 0x5, 0x5, 0xbef509fecafaee8a)
	/usr/local/go/src/testing/testing.go:1061 +0x2c4
testing.(*M).Run(0xc4205d1180, 0x0)
	/usr/local/go/src/testing/testing.go:978 +0x171
github.com/cockroachdb/cockroach/pkg/sql/logictest.TestMain(0xc4205d1180)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/main_test.go:37 +0xb8
main.main()
	_testmain.go:48 +0x151
```

--- a/pkg/sql/opt/ordering/ordering.go
+++ b/pkg/sql/opt/ordering/ordering.go
@@ -251,6 +251,9 @@ func remapProvided(provided opt.Ordering, fds *props.FuncDepSet, outCols opt.Col
 func trimProvided(
 	provided opt.Ordering, required *physical.OrderingChoice, fds *props.FuncDepSet,
 ) opt.Ordering {
+	if len(provided) == 0 {
+		return nil
+	}
 	// closure is the set of columns that are functionally determined by the
 	// columns in provided[:provIdx].
 	closure := fds.ComputeClosure(opt.ColSet{})
@@ -260,12 +263,12 @@ func trimProvided(
 		// Consume columns from the provided ordering until their closure intersects
 		// the required group.
 		for !closure.Intersects(c.Group) {
-			if provIdx == len(provided) {
-				panic("provided does not intersect required")
-			}
 			closure.Add(int(provided[provIdx].ID()))
 			closure = fds.ComputeClosure(closure)
 			provIdx++
+			if provIdx == len(provided) {
+				return provided
+			}
 		}
 	}
 	return provided[:provIdx]
@@ -301,21 +304,28 @@ func checkProvided(expr memo.RelExpr, required *physical.OrderingChoice, provide
 		))
 	}
 
-	// The provided ordering must intersect the required ordering, after FDs are
-	// applied.
-	fds := &expr.Relational().FuncDeps
-	r := required.Copy()
-	r.Simplify(fds)
-	var p physical.OrderingChoice
-	p.FromOrdering(provided)
-	p.Simplify(fds)
-	if !r.Any() && (p.Any() || !p.Intersects(&r)) {
-		panic(fmt.Sprintf(
-			"provided %s does not intersect required %s (FDs: %s)", provided, required, fds,
-		))
+	// TODO(radu): this check would be nice to have, but it is too strict. In some
+	// cases, child expressions created during exploration (like constrained
+	// scans) have FDs that are more restricted than what was known when the
+	// parent expression was constructed. Related to #32320.
+	if false {
+		// The provided ordering must intersect the required ordering, after FDs are
+		// applied.
+		fds := &expr.Relational().FuncDeps
+		r := required.Copy()
+		r.Simplify(fds)
+		var p physical.OrderingChoice
+		p.FromOrdering(provided)
+		p.Simplify(fds)
+		if !r.Any() && (p.Any() || !p.Intersects(&r)) {
+			panic(fmt.Sprintf(
+				"provided %s does not intersect required %s (FDs: %s)", provided, required, fds,
+			))
+		}
 	}
 
 	// The provided ordering should not have unnecessary columns.
+	fds := &expr.Relational().FuncDeps
 	if trimmed := trimProvided(provided, required, fds); len(trimmed) != len(provided) {
 		panic(fmt.Sprintf("provided %s can be trimmed to %s (FDs: %s)", provided, trimmed, fds))
 	}

